modules/extension/app-schema/app-schema/src/main/java/org/geotools/data/complex/filter/FeatureChainedAttributeVisitor.java

```diff
@@ -0,0 +1,573 @@
+                            + "against was provided: nested feature type cannot be determined");
+                    // quit the search
+                    return;
+                } else {
+                    FeatureTypeMapping nestedType = nestedAttr.getFeatureTypeMapping(feature);
+                    if (nestedType != null) {
+                        AttributeType nestedPropertyType = nestedType.getTargetFeature().getType();
+                        QName nestedTypeQName = getFeatureTypeQName(nestedType);
+                        Step nestedTypeStep = new Step(nestedTypeQName, 1);
+                        StepList nestedTypeXPath = targetXPath.clone();
+                        nestedTypeXPath.add(nestedTypeStep);
+
+                        boolean xpathContainsNestedType = currentXPath.startsWith(nestedTypeXPath);
+                        boolean hasSimpleContent = Types.isSimpleContentType(nestedPropertyType);
+
+                        // if this is feature chaining for simple content, the name of the nested type
+                        // may not be present in the XPath, as it was already specified as the container
+                        // property (e.g. see mappings doing chaining for gml:name)
+                        if (xpathContainsNestedType || hasSimpleContent) {
+                            LOGGER.finer("Nested feature type found: " + nestedTypeQName);
+                            FeatureChainedAttributeDescriptor copy = attrDescr.shallowCopy();
+                            copy.addLink(new FeatureChainLink(currentType, nestedAttr));
+
+                            // new root mapping to search
+                            FeatureTypeMapping newType = nestedType;
+
+                            // new xpath
+                            StepList newXPath = currentXPath.clone();
+                            int startIdx = (xpathContainsNestedType) ? nestedTypeXPath.size()
+                                    : currentXPath.size();
+                            newXPath = newXPath.subList(startIdx, currentXPath.size());
+
+                            // if nested type has simple content, XPath expression may point directly
+                            // to the type, and not to one of its attributes (which, BTW, can only
+                            // be client properties, or it wouldn't be simple content)
+                            if (newXPath.isEmpty() && hasSimpleContent) {
+                                newXPath.add(nestedTypeStep);
+                            }
+
+                            // recursive call
+                            walkXPathRecursive(newXPath, newType, copy, feature);
+                            // I'm not done yet
+                            searchIsOver = false;
+                        }
+                    } else {
+                        LOGGER.fine("Nested feature type could not be determined");
```

Comment:
Perhaps include some more information here to aid debugging? The xpath, perhaps?